File: packages/server/cmd/renewlet/ai_recognition_diagnostics.go
package main

// ai_recognition_diagnostics.go 构造 AI 识别的临时排障信息。
//
// diagnostics 只随当前认证 API 响应返回，不入库、不导出；prompt、raw 输出、provider metadata 和图片信息
// 必须在这里统一脱敏/截断，避免第三方 key、Bearer、JWT、cookie 或 base64 图片泄漏到浏览器日志。
import (
	"encoding/json"
	"errors"
	"strings"
)

const (
	aiRecognitionDiagnosticSchemaVersion = "1"
	aiRecognitionDiagnosticTextMaxChars  = 32000
	aiRecognitionDiagnosticJSONMaxChars  = 32000
)

type aiRecognitionDiagnosticText struct {
	Value     string `json:"value"`
	Truncated bool   `json:"truncated"`
}

type aiRecognitionDiagnostics struct {
	SchemaVersion string                          `json:"schemaVersion"`
	PromptVersion string                          `json:"promptVersion"`
	SchemaName    string                          `json:"schemaName"`
	Prompt        aiRecognitionDiagnosticPrompt   `json:"prompt"`
	Output        aiRecognitionDiagnosticOutput   `json:"output"`
	Request       aiRecognitionDiagnosticRequest  `json:"request"`
	Response      aiRecognitionDiagnosticResponse `json:"response"`
}

type aiRecognitionDiagnosticPrompt struct {
	System aiRecognitionDiagnosticText `json:"system"`
	User   aiRecognitionDiagnosticText `json:"user"`
}

type aiRecognitionDiagnosticOutput struct {
	RawModelText  *aiRecognitionDiagnosticText `json:"rawModelText"`
	RawObjectJSON *aiRecognitionDiagnosticText `json:"rawObjectJson"`
}

type aiRecognitionDiagnosticRequest struct {
	ProviderType      string                         `json:"providerType"`
	TransportProtocol string                         `json:"transportProtocol"`
	Model             string                         `json:"model"`
	ThinkingControl   *aiThinkingControl             `json:"thinkingControl"`
	MaxOutputTokens   int                            `json:"maxOutputTokens"`
	TextCharCount     int                            `json:"textCharCount"`
	Images            []aiRecognitionDiagnosticImage `json:"images"`
}

type aiRecognitionDiagnosticImage struct {
	MediaType string `json:"mediaType"`
	SizeBytes int    `json:"sizeBytes"`
}

type aiRecognitionDiagnosticResponse struct {
	Usage            interface{} `json:"usage"`
	FinishReason     *string     `json:"finishReason"`
	ProviderMetadata interface{} `json:"providerMetadata"`
}

type aiRecognitionCapture struct {
	rawModelText     string
	usage            interface{}
	finishReason     string
	providerMetadata interface{}
}

type aiRecognitionRunError struct {
	cause       error
	diagnostics aiRecognitionDiagnostics
}

func (err *aiRecognitionRunError) Error() string {
	if err == nil || err.cause == nil {
		return "AI_RECOGNITION_FAILED"
	}
	return err.cause.Error()
}

func (err *aiRecognitionRunError) Unwrap() error {
	if err == nil {
		return nil
	}
	return err.cause
}

func buildAIRecognitionDiagnostics(settings aiRecognitionSettings, input aiRecognitionInput, systemPrompt string, userPrompt string, rawModelText string, rawObject interface{}, usage interface{}, finishReason string, providerMetadata interface{}) aiRecognitionDiagnostics {
	images := make([]aiRecognitionDiagnosticImage, 0, len(input.Images))
	for _, image := range input.Images {
		// 图片诊断只暴露类型和大小，绝不返回 data URL/base64；用户可据此排查输入规模而不泄漏内容。
		images = append(images, aiRecognitionDiagnosticImage{MediaType: image.MediaType, SizeBytes: image.SizeBytes})
	}
	diagnostics := aiRecognitionDiagnostics{
		SchemaVersion: aiRecognitionDiagnosticSchemaVersion,
		PromptVersion: aiRecognitionPrompt.Version,
		SchemaName:    aiRecognitionPrompt.SchemaName,
		Prompt: aiRecognitionDiagnosticPrompt{
			System: diagnosticAIRecognitionText(systemPrompt, aiRecognitionDiagnosticTextMaxChars),
			User:   diagnosticAIRecognitionText(userPrompt, aiRecognitionDiagnosticTextMaxChars),
		},
		Output: aiRecognitionDiagnosticOutput{
			RawModelText:  diagnosticAIRecognitionOptionalText(rawModelText, aiRecognitionDiagnosticTextMaxChars),
			RawObjectJSON: diagnosticAIRecognitionOptionalText(aiRecognitionJSONText(rawObject), aiRecognitionDiagnosticJSONMaxChars),
		},
		Request: aiRecognitionDiagnosticRequest{
			ProviderType:      settings.ProviderType,
			TransportProtocol: settings.TransportProtocol,
			Model:             settings.Model,
			ThinkingControl:   input.ThinkingControl,
			MaxOutputTokens:   aiRecognitionOutputTokenLimit(input),
			TextCharCount:     len([]rune(input.Text)),
			Images:            images,
		},
		Response: aiRecognitionDiagnosticResponse{
			Usage:            sanitizeAIRecognitionDiagnosticJSON(usage),
			FinishReason:     optionalAIRecognitionString(finishReason),
			ProviderMetadata: sanitizeAIRecognitionDiagnosticJSON(providerMetadata),
		},
	}
	return diagnostics
}

func diagnosticAIRecognitionOptionalText(value string, maxChars int) *aiRecognitionDiagnosticText {
	if value == "" {
		return nil
	}
	text := diagnosticAIRecognitionText(value, maxChars)
	return &text
}

func diagnosticAIRecognitionText(value string, maxChars int) aiRecognitionDiagnosticText {
	safe := redactAIRecognitionSecrets(value)
	runes := []rune(safe)
	if len(runes) > maxChars {
		return aiRecognitionDiagnosticText{Value: string(runes[:maxChars]), Truncated: true}
	}
	return aiRecognitionDiagnosticText{Value: safe, Truncated: false}
}

func aiRecognitionJSONText(value interface{}) string {
	if value == nil {
		return ""
	}
	data, err := json.MarshalIndent(value, "", "  ")
	if err != nil {
		return err.Error()
	}
	return string(data)
}

func sanitizeAIRecognitionDiagnosticJSON(value interface{}) interface{} {
	if value == nil {
		return nil
	}
	text := diagnosticAIRecognitionText(aiRecognitionJSONText(value), aiRecognitionDiagnosticJSONMaxChars)
	if text.Truncated {
		// JSON 超限时保持文本形态，避免截断后的半截 JSON 被误解析成可信 provider metadata。
		return text
	}
	var parsed interface{}
	if err := json.Unmarshal([]byte(text.Value), &parsed); err != nil {
		return text
	}
	return parsed
}

func optionalAIRecognitionString(value string) *string {
	value = strings.TrimSpace(value)
	if value == "" {
		return nil
	}
	return &value
}

func aiRecognitionDiagnosticsFromError(err error) *aiRecognitionDiagnostics {
	var runErr *aiRecognitionRunError
	if errors.As(err, &runErr) {
		return &runErr.diagnostics
	}
	return nil
}

func aiRecognitionCauseError(err error) error {
	var runErr *aiRecognitionRunError
	if errors.As(err, &runErr) {
		return runErr.cause
	}
	return err
}

func aiRecognitionRawTextFromError(err error) string {
	if err == nil {
		return ""
	}
	message := err.Error()
	const marker = "(raw: "
	index := strings.LastIndex(message, marker)
	if index < 0 {
		return ""
	}
	raw := message[index+len(marker):]
	return strings.TrimSuffix(raw, ")")
}

File: packages/cloudflare/src/ai-recognition-diagnostics.ts
/**
 * Cloudflare AI 识别 diagnostics 构造与脱敏。
 *
 * diagnostics 只允许随当前认证响应返回；prompt、raw 输出、usage/provider metadata 必须在这里截断和脱敏。
 */
import { NoObjectGeneratedError } from "ai";
import {
  AI_RECOGNITION_DIAGNOSTIC_JSON_MAX_CHARS,
  AI_RECOGNITION_DIAGNOSTIC_TEXT_MAX_CHARS,
  aiRecognitionDiagnosticsSchema,
  aiRecognitionErrorDetailsSchema,
  type AiRecognitionDiagnostics,
  type AiRecognitionSettings,
  type AiProviderResponse,
  type AiThinkingControl,
} from "@renewlet/shared/schemas/ai-recognition";
import {
  AI_RECOGNITION_PROMPT_VERSION,
  AI_RECOGNITION_SCHEMA_NAME,
} from "@renewlet/shared/ai-recognition-prompt";
import { providerResponseFromError } from "./ai-provider-response";

const AI_SECRET_PATTERN = /(sk-[A-Za-z0-9_-]{8,}|AIza[0-9A-Za-z_-]{8,}|sk-ant-[A-Za-z0-9_-]{8,}|Bearer\s+[A-Za-z0-9._-]{8,}|eyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+|(?:api[_-]?key|authorization|cookie|set-cookie|access[_-]?token|refresh[_-]?token)["'\s:=]+[A-Za-z0-9._~+/=-]{8,})/gi;

interface AIRecognitionDiagnosticsInput {
  text: string;
  images: Array<{ mediaType: string; data: Uint8Array }>;
}

export function buildAIRecognitionDiagnostics({
  settings,
  input,
  thinkingControl,
  maxOutputTokens,
  systemPrompt,
  userPrompt,
  rawModelText,
  rawObject,
  usage,
  finishReason,
  providerMetadata,
}: {
  settings: AiRecognitionSettings;
  input: AIRecognitionDiagnosticsInput;
  thinkingControl: AiThinkingControl | null;
  maxOutputTokens: number;
  systemPrompt: string;
  userPrompt: string;
  rawModelText: string | null;
  rawObject: unknown;
  usage: unknown;
  finishReason: string | null;
  providerMetadata: unknown;
}): AiRecognitionDiagnostics {
  // diagnostics 只进入当前 API 响应，不能入库；这里集中截断/脱敏，避免模型平台原文泄漏密钥。
  return aiRecognitionDiagnosticsSchema.parse({
    schemaVersion: "1",
    promptVersion: AI_RECOGNITION_PROMPT_VERSION,
    schemaName: AI_RECOGNITION_SCHEMA_NAME,
    prompt: {
      system: diagnosticText(systemPrompt, AI_RECOGNITION_DIAGNOSTIC_TEXT_MAX_CHARS),
      user: diagnosticText(userPrompt, AI_RECOGNITION_DIAGNOSTIC_TEXT_MAX_CHARS),
    },
    output: {
      rawModelText: rawModelText === null ? null : diagnosticText(rawModelText, AI_RECOGNITION_DIAGNOSTIC_TEXT_MAX_CHARS),
      rawObjectJson: rawObject === null ? null : diagnosticText(safeJsonStringify(rawObject), AI_RECOGNITION_DIAGNOSTIC_JSON_MAX_CHARS),
    },
    request: {
      providerType: settings.providerType,
      transportProtocol: settings.transportProtocol,
      model: settings.model,
      thinkingControl,
      maxOutputTokens,
      textCharCount: [...input.text].length,
      // 图片诊断只暴露类型和大小，绝不返回 bytes/data URL，避免浏览器日志保存用户上传原图。
      images: input.images.map((image) => ({ mediaType: image.mediaType, sizeBytes: image.data.byteLength })),
    },
    response: {
      usage: sanitizeDiagnosticJson(usage),
      finishReason,
      providerMetadata: sanitizeDiagnosticJson(providerMetadata),
    },
  });
}

export function safeAIRecognitionError(error: unknown): string {
  return redactAIRecognitionSecrets(error instanceof Error ? error.message : String(error)).slice(0, 500);
}

export function aiRecognitionErrorDetails(
  reason: string,
  error: unknown,
  diagnostics: AiRecognitionDiagnostics,
  providerResponse: AiProviderResponse | null = providerResponseFromError(error),
) {
  return aiRecognitionErrorDetailsSchema.parse({
    reason,
    providerMessage: error === null ? null : safeAIRecognitionError(error),
    providerResponse,
    diagnostics,
  });
}

export function noObjectGeneratedText(error: unknown): string | null {
  return NoObjectGeneratedError.isInstance(error) && typeof error.text === "string" ? error.text : null;
}

export function noObjectGeneratedUsage(error: unknown): unknown | null {
  return NoObjectGeneratedError.isInstance(error) ? error.usage ?? null : null;
}

export function noObjectGeneratedFinishReason(error: unknown): string | null {
  return NoObjectGeneratedError.isInstance(error) ? finishReasonText(error.finishReason) : null;
}

export function finishReasonText(value: unknown): string | null {
  if (typeof value === "string" && value.trim()) return value;
  if (isDiagnosticRecord(value)) {
    const unified = value["unified"];
    if (typeof unified === "string" && unified.trim()) return unified;
    const raw = value["raw"];
    if (typeof raw === "string" && raw.trim()) return raw;
  }
  return null;
}

export function extractAIModelText(content: unknown): string | null {
  if (!Array.isArray(content)) return null;
  const text = content.map((part) => {
    if (!isDiagnosticRecord(part) || part["type"] !== "text") return "";
    const value = part["text"];
    return typeof value === "string" ? value : "";
  }).join("");
  return text || null;
}

function diagnosticText(value: string, maxChars: number) {
  const safe = redactAIRecognitionSecrets(value);
  const chars = [...safe];
  return {
    value: chars.slice(0, maxChars).join(""),
    truncated: chars.length > maxChars,
  };
}

function sanitizeDiagnosticJson(value: unknown): unknown | null {
  if (value === undefined || value === null) return null;
  const text = diagnosticText(safeJsonStringify(value), AI_RECOGNITION_DIAGNOSTIC_JSON_MAX_CHARS);
  if (text.truncated) return text;
  try {
    return JSON.parse(text.value) as unknown;
  } catch {
    return text;
  }
}

function safeJsonStringify(value: unknown): string {
  const seen = new WeakSet<object>();
  try {
    return JSON.stringify(value, (_key, item: unknown) => {
      if (typeof item === "bigint") return item.toString();
      if (item && typeof item === "object") {
        if (seen.has(item)) return "[Circular]";
        seen.add(item);
      }
      return item;
    }, 2);
  } catch (error) {
    return error instanceof Error ? error.message : String(error);
  }
}

function redactAIRecognitionSecrets(value: string): string {
  return value.replace(AI_SECRET_PATTERN, "[redacted]");
}

function isDiagnosticRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}
